src/main/scala/tilelink/Bundles.scala

```diff
@@ -134,8 +137,59 @@ final class TLBundleA(params: TLBundleParameters)
   val mask    = UInt(width = params.dataBits/8)
   val data    = UInt(width = params.dataBits)
   val corrupt = Bool() // only applies to *Data messages
-}
 
+  def opcodePrint()
+  {
+    switch (opcode) {
+      is(0.U) {printf("OP: PutFullData      ")} 
+      is(1.U) {printf("OP: PutPartialData   ")}
+      is(2.U) {printf("OP: ArithmeticData   ")} 
+      is(3.U) {printf("OP: LogicalData      ")} 
+      is(4.U) {printf("OP: Get              ")}
+      is(5.U) {printf("OP: Hint             ")}
+      is(6.U) {printf("OP: AcquireBlock     ")}
+      is(7.U) {printf("OP: AcquirePerm      ")}
+    }
+  }
```

Comment:
A functional style can help with the clutter. One equivalent approach would be:
```scala
  def opcodePrint(): Unit =
    Seq("PutFullData", "PutPartialData", "ArithmeticData", "LogicalData", "Get", "Hint", "AcquireBlock", "AcquirePerm")
      .zipWithIndex
      .map{ case (str, idx) => when (opcode === idx.U) { printf(s"OP: $str\n") } }
```

Reply:
Can this be done in runtime, because thats where we will be using it.

Reply:
Both methods will statically emit circuits with `printf`s that trip on the conditions specified. So, they will generate hardware `printf`s that are usable at runtime.

Note: there is a subtle difference here in that `switch`/`is` produces a `when`/`else` nesting while the explicit `when` blocks of the functional style are all flat. Functionally, there's no difference, though.

The original is doing something like:
```
    node _T_33 = eq(UInt<1>("h00"), UInt<1>("h01")) 
    when _T_33 : 
      node _T_34 = bits(reset, 0, 0) 
      node _T_35 = eq(_T_34, UInt<1>("h00")) 
      when _T_35 : 
        printf(clock, UInt<1>(1), "OP: PutFullData      ") 
    else : 
      node _T_36 = eq(UInt<1>("h01"), UInt<1>("h01")) 
      when _T_36 : 
        node _T_37 = bits(reset, 0, 0) 
        node _T_38 = eq(_T_37, UInt<1>("h00")) 
        when _T_38 : 
          printf(clock, UInt<1>(1), "OP: PutPartialData   ") 
```

The `map` is generating:
```
    node _T_9 = eq(UInt<1>("h01"), UInt<1>("h00")) 
    when _T_9 : 
      node _T_10 = bits(reset, 0, 0) 
      node _T_11 = eq(_T_10, UInt<1>("h00")) 
      when _T_11 : 
        printf(clock, UInt<1>(1), "OP: PutFullData\n") 
    node _T_12 = eq(UInt<1>("h01"), UInt<1>("h01")) 
    when _T_12 : 
      node _T_13 = bits(reset, 0, 0) 
      node _T_14 = eq(_T_13, UInt<1>("h00")) 
      when _T_14 : 
        printf(clock, UInt<1>(1), "OP: PutPartialData\n") 
```